File: packages/core/src/db/commits.ts
import type { Database } from "bun:sqlite";
import type { CardId, ExecutionId, FileChange } from "../types/index.js";

export interface CommitId extends String {
  readonly __brand: "CommitId";
}

export interface CardCommit {
  id: string;
  card_id: string;
  execution_id: string | null;
  sha: string;
  message: string;
  author_name: string;
  author_email: string;
  files_changed: string | null;
  created_at: string;
}

export interface CreateCommitInput {
  sha: string;
  message: string;
  authorName: string;
  authorEmail: string;
  filesChanged: FileChange[];
}

export function listCommits(
  db: Database,
  cardId: CardId
): CardCommit[] {
  return db
    .query(
      "SELECT * FROM card_commits WHERE card_id = ? ORDER BY created_at DESC"
    )
    .all(cardId) as CardCommit[];
}

export function addCommit(
  db: Database,
  cardId: CardId,
  executionId: ExecutionId | null,
  input: CreateCommitInput
): CardCommit {
  const row = db
    .query(
      `INSERT INTO card_commits (card_id, execution_id, sha, message, author_name, author_email, files_changed)
       VALUES (?, ?, ?, ?, ?, ?, ?)
       RETURNING *`
    )
    .get(
      cardId,
      executionId,
      input.sha,
      input.message,
      input.authorName,
      input.authorEmail,
      JSON.stringify(input.filesChanged)
    ) as CardCommit;
  return row;
}

export function addCommits(
  db: Database,
  cardId: CardId,
  executionId: ExecutionId | null,
  inputs: CreateCommitInput[]
): CardCommit[] {
  const results: CardCommit[] = [];
  const stmt = db.query(
    `INSERT INTO card_commits (card_id, execution_id, sha, message, author_name, author_email, files_changed)
     VALUES (?, ?, ?, ?, ?, ?, ?)
     RETURNING *`
  );
  const tx = db.transaction(() => {
    for (const input of inputs) {
      const row = stmt.get(
        cardId,
        executionId,
        input.sha,
        input.message,
        input.authorName,
        input.authorEmail,
        JSON.stringify(input.filesChanged)
      ) as CardCommit;
      results.push(row);
    }
  });
  tx();
  return results;
}

File: rust/crates/core/src/db/commits.rs
use rusqlite::{params, Connection};

use crate::types::CardCommit;
use crate::Result;

pub struct NewCommit<'a> {
    pub card_id: &'a str,
    pub execution_id: Option<&'a str>,
    pub sha: &'a str,
    pub message: &'a str,
    pub author_name: &'a str,
    pub author_email: &'a str,
    pub files_changed: Option<&'a str>,
}

pub fn record(conn: &Connection, c: &NewCommit) -> Result<CardCommit> {
    let row = conn.query_row(
        "INSERT INTO card_commits (card_id, execution_id, sha, message, author_name, author_email, files_changed)
         VALUES (?, ?, ?, ?, ?, ?, ?)
         RETURNING *",
        params![c.card_id, c.execution_id, c.sha, c.message, c.author_name, c.author_email, c.files_changed],
        row_to_commit,
    )?;
    Ok(row)
}

pub fn list_for_card(conn: &Connection, card_id: &str) -> Result<Vec<CardCommit>> {
    let mut stmt = conn.prepare(
        "SELECT * FROM card_commits WHERE card_id = ? ORDER BY created_at DESC",
    )?;
    let rows = stmt
        .query_map([card_id], row_to_commit)?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(rows)
}

fn row_to_commit(row: &rusqlite::Row<'_>) -> rusqlite::Result<CardCommit> {
    Ok(CardCommit {
        id: row.get("id")?,
        card_id: row.get("card_id")?,
        execution_id: row.get("execution_id")?,
        sha: row.get("sha")?,
        message: row.get("message")?,
        author_name: row.get("author_name")?,
        author_email: row.get("author_email")?,
        files_changed: row.get("files_changed")?,
        created_at: row.get("created_at")?,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::{boards, cards, connection::open_memory};
    use crate::types::{CreateBoard, CreateCard};

    #[test]
    fn record_then_list() {
        let conn = open_memory().unwrap();
        let b = boards::create(&conn, &CreateBoard {
            name: "b".into(), description: "".into(), directory: "/tmp".into(),
            color: None, slug: None, github_url: None,
        }).unwrap();
        let c = cards::create(&conn, &b.id, &CreateCard { title: "c".into(), description: None, tags: None, assignee: None }).unwrap();
        record(&conn, &NewCommit {
            card_id: &c.card.id,
            execution_id: None,
            sha: "abc123",
            message: "wip",
            author_name: "Tester",
            author_email: "test@example.com",
            files_changed: Some("a.txt"),
        }).unwrap();
        let list = list_for_card(&conn, &c.card.id).unwrap();
        assert_eq!(list.len(), 1);
        assert_eq!(list[0].sha, "abc123");
    }
}
